README.md
# knob

A simple knob UI widget.

## Getting Started

Add the dependency to your `pubspec.yaml`:

```
dependencies:
  flutter:
    sdk: flutter
  flutter_knob: // Add it here
```



This package provides a Knob widget with a similar API as Flutter's Slider widget.

Working example, showing the Knob controlling, and being controlled by a Slider:

<img src="./knob-and-slider.png" width="300"/>

```
import 'package:flutter/material.dart';

import 'package:flutter_knob/flutter_knob.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  double _value = 0.0;
  void _setValue(double value) => setState(() => _value = value);
  static const double minValue = 0;
  static const double maxValue = 10;

  @override
  Widget build(BuildContext context) {

    return Scaffold(
      appBar: AppBar(
        title: Text('Flutter Tutorial'),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Knob(
              value: _value,
              color: Colors.blue,
              onChanged: _setValue,
              min: minValue,
              max: maxValue,
              size: 50
            ),
            Slider(
                value: _value,
                onChanged: _setValue,
                min: minValue,
                max: maxValue),
            Text(
              'Value: ${_value.toStringAsFixed(3)}',
            ),
          ],
        ),
      ),
    );
  }
}
```



CHANGELOG.md
## [0.0.1] - 10/08/2019

A very basic knob for use in Flutter, with a similar interfce to the Slider widget


lib/knob.dart
library knob;

import 'package:flutter/material.dart';
import 'dart:math';

class Knob extends StatefulWidget {

  // Define the parameters of this widget
  final double value;
  final double min;
  final double max;

  // Two extra parameters to make the widget more easy to customise
  final double size;
  final Color color;

  // ValueChanged is a type built into Dart for a function that changes a value
  final ValueChanged<double> onChanged;

  // Define a build method for the widget which uses these parameters
  Knob({this.value, this.min = 0, this.max = 1, this.color = Colors.blue, this.size = 50, this.onChanged});

  @override
  State<StatefulWidget> createState() => KnobState();
}

class KnobState extends State<Knob> {

  // These are static constants because they are in internal parameters of the knob that
  // can't be changed from the outside
  static const double minAngle = -160;
  static const double maxAngle = 160;
  static const double sweepAngle = maxAngle - minAngle;
  
  @override
  Widget build(BuildContext context) {
    // distanceToAngle now depends on potentially non-constant parameters,
    // so must be defined in the build method
    double distanceToAngle = 0.007 * (widget.max - widget.min);

    double normalisedValue = (widget.value - widget.min)/(widget.max - widget.min);
    double angle = (minAngle + normalisedValue * sweepAngle) * 2 * pi / 360;
    double size = widget.size;
    return Center(
      child: Container(
        width: size,
        height: size,
        child: Transform.rotate(
          angle: angle,
          child: GestureDetector(
            onVerticalDragUpdate: (DragUpdateDetails details) {
              double changeInY = -details.delta.dy;
              double changeInValue = distanceToAngle * changeInY;
              double newValue = widget.value + changeInValue;
              double clippedValue = min(max(newValue, widget.min), widget.max);

              widget.onChanged(clippedValue);
            },
            child: ClipOval(
              child: Container(
                color: widget.color,
                child: Icon(
                  Icons.arrow_upward,
                  color: Colors.white,
                  size: size,
                )
              )
            ),
          ),
        ),
      ),
    );
  }
}

lib/flutter_knob.dart
library flutter_knob;

import 'package:flutter/material.dart';
import 'dart:math';

class Knob extends StatefulWidget {

  // Define the parameters of this widget
  final double value;
  final double min;
  final double max;

  // Two extra parameters to make the widget more easy to customise
  final double size;
  final Color color;

  // ValueChanged is a type built into Dart for a function that changes a value
  final ValueChanged<double> onChanged;

  // Define a build method for the widget which uses these parameters
  Knob({this.value, this.min = 0, this.max = 1, this.color = Colors.blue, this.size = 50, this.onChanged});

  @override
  State<StatefulWidget> createState() => KnobState();
}

class KnobState extends State<Knob> {

  // These are static constants because they are in internal parameters of the knob that
  // can't be changed from the outside
  static const double minAngle = -160;
  static const double maxAngle = 160;
  static const double sweepAngle = maxAngle - minAngle;
  
  @override
  Widget build(BuildContext context) {
    // distanceToAngle now depends on potentially non-constant parameters,
    // so must be defined in the build method
    double distanceToAngle = 0.007 * (widget.max - widget.min);

    double normalisedValue = (widget.value - widget.min)/(widget.max - widget.min);
    double angle = (minAngle + normalisedValue * sweepAngle) * 2 * pi / 360;
    double size = widget.size;
    return Center(
      child: Container(
        width: size,
        height: size,
        child: Transform.rotate(
          angle: angle,
          child: GestureDetector(
            onVerticalDragUpdate: (DragUpdateDetails details) {
              double changeInY = -details.delta.dy;
              double changeInValue = distanceToAngle * changeInY;
              double newValue = widget.value + changeInValue;
              double clippedValue = min(max(newValue, widget.min), widget.max);

              widget.onChanged(clippedValue);
            },
            child: ClipOval(
              child: Container(
                color: widget.color,
                child: Icon(
                  Icons.arrow_upward,
                  color: Colors.white,
                  size: size,
                )
              )
            ),
          ),
        ),
      ),
    );
  }
}